sdks/go/pkg/beam/io/textio/textio.go

```diff
@@ -135,55 +134,53 @@ func Write(s beam.Scope, filename string, col beam.PCollection) {
 	s = s.Scope("textio.Write")
 
 	validateScheme(filename)
-	beam.ParDo0(s, &writeFileFn{Filename: filename}, col)
+
+	// NOTE(BEAM-3579): We may never call Teardown for non-local runners and
```

Comment:
Teardown happens beyond the unit of commit, which is why it's not appropriate here, as far as I can tell - performing anything that 'must be done' in anything other than start/finish bundle and processElement is broken because you can arbitrarily lose data that can never be retried

Reply:
Yes. That was the bug here. On Dataflow, the lack of an orderly shutdown means that we never call Teardown even in the successful case where no failures happened.

Reply:
What I mean is that it's not a data-relative function - so buffering elements in teardown in a production runner that guarantees an orderly shutdown is still not appropriate, because if the worker dies the elements that produced the contents of the buffer will never be replayed, and the data is lost. Teardown is never an appropriate location to buffer things that you're not ok with losing. This isn't documented strongly enough in the Java SDK, at minimum, but should be.

Reply:
I understand. I was unclear before. I meant: "In addition, on Dataflow, ...". A better-defined semantics for when Teardown is desirable, but would not address the issue in the PR regardless.